File: Python/121.Best Time to Buy and Sell Stock.py
#一用一个数组表示股票每天的价格，数组的第i个数表示股票在第i天的价格。 如果只允许进行一次交易，也就是说只允许买一支股票并卖掉，求最大的收益
class Solution(object):
    def maxProfit(self, prices):
        l  = len(prices)
        if l == 0:
            return 0
        maxpro , curmin = 0 , prices[0]
        for i in range(1,l):
            maxpro = max(maxpro,prices[i]-curmin)
            curmin = min(prices[i],curmin)
        return maxpro


s = Solution()
print(s.maxProfit([84,22,3,44,2]))

File: C++/Dynamic Programming/121.Best Time to Buy and Sell Stock.cpp
/*
最佳的购买和抛售股票的时机，最多买卖各一次，求最大收益。
*/
/*
思路：记录前面最小值，和最大落差。
 */
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
    	if(prices.size()==0)    return 0;
        int maxpro = 0 , curmin = prices[0];
        for(int i=1; i < prices.size(); ++i)
        {
        	maxpro = max(maxpro,prices[i]-curmin);
        	curmin = min(curmin,prices[i]);
        }
        return maxpro;
    }
};
int main()
{
	Solution s;
	vector<int> prices{7, 1, 5, 3, 6, 4};
	cout<<s.maxProfit(prices)<<endl;
}

